namui-animation-editor/src/types/history_system/history.rs

```diff
@@ -0,0 +1,109 @@
+pub struct History<TState> {
+    undo_count: usize,
+    initial_state: TState,
+    states: Vec<TState>,
+}
+
+impl<TState> History<TState> {
+    pub fn new(initial_state: TState) -> Self {
+        Self {
+            undo_count: 0,
+            initial_state,
+            states: Vec::new(),
+        }
+    }
+    pub fn push(&mut self, state: TState) {
+        for _ in 0..(self.undo_count.min(self.states.len())) {
+            self.states.pop();
+        }
+        self.undo_count = 0;
+        self.states.push(state);
+    }
+    pub fn undo(&mut self) -> Option<()> {
+        if self.undo_count < self.states.len() {
+            self.undo_count += 1;
+            Some(())
+        } else {
+            None
+        }
+    }
+    pub fn get(&self) -> &TState {
+        if self.states.len() <= self.undo_count {
+            &self.initial_state
+        } else {
+            &self.states[self.states.len() - self.undo_count - 1]
+        }
+    }
+    pub fn redo(&mut self) -> Option<()> {
+        if self.undo_count > 0 {
+            self.undo_count -= 1;
+            Some(())
+        } else {
+            None
+        }
+    }
```

Comment:
Not `Result` but `Option`. Is it intended?

Reply:
No, seems bad. let me fix it.